base/broadcast.jl

```diff
@@ -257,13 +257,23 @@ end
 @inline broadcast_elwise_op(f, As...) =
     broadcast!(f, similar(Array{promote_eltype_op(f, As...)}, broadcast_indices(As...)), As...)
 
-@pure typestuple(a) = Tuple{eltype(a)}
-@pure typestuple(T::Type) = Tuple{Type{T}}
-@pure typestuple(a, b...) = Tuple{typestuple(a).types..., typestuple(b...).types...}
+typestuple(a) = (Base.@_pure_meta; Tuple{eltype(a)})
+typestuple(T::Type) = (Base.@_pure_meta; Tuple{Type{T}})
+typestuple(a, b...) = (Base.@_pure_meta; Tuple{typestuple(a).types..., typestuple(b...).types...})
+
+ftype(f, A) = typeof(a->f(a))
+ftype(f, A...) = typeof(a->f(a...))
+ftype(T::Type, A...) = Type{T}
+ziptype(A) = (Base.@_pure_meta; Tuple{eltype(A)})
+ziptype(A, B) = (Base.@_pure_meta; Iterators.Zip2{ziptype(A), ziptype(B)})
+@inline ziptype(A, B, C, D...) = Iterators.Zip{ziptype(A),ziptype(B, C, D...)}
+
+_broadcast_type(f, T::Type, As...) = Base._return_type(f, typestuple(T, As...))
```

Comment:
I imagine this construction handles the test case for #19421, but will not handle types in other argument positions?